src/fsharp/FSharp.Core/async.fs

```diff
@@ -1250,31 +1250,28 @@ namespace Microsoft.FSharp.Control
                                 |> unfake)
                     | Some maxDegreeOfParallelism ->
                         let mutable i = -1
-                        let worker = MakeAsync (fun _ ->
-                            while i < tasks.Length do
+                        let rec worker (trampolineHolder : TrampolineHolder) =
+                            if i < tasks.Length then
                                 let j = Interlocked.Increment &i
                                 if j < tasks.Length then
-                                    let trampolineHolder = new TrampolineHolder()
-                                    trampolineHolder.ExecuteWithTrampoline (fun () ->
-                                        let ctxt =
+                                    if innerCTS.Token.IsCancellationRequested then
+                                        let cexn = new OperationCanceledException (innerCTS.Token)
+                                        recordFailure (Choice2Of2 cexn) |> unfake
+                                        worker trampolineHolder |> unfake
```

Comment:
I don't think you need to iterate through the remaining tasks if innerCts has been triggered.

Reply:
recordFailure won't trigger the final continuation until it's been called tasks.Length times. There might be a performance improvement here but it's not as simple as just immediately calling the final continuation (there might be other workers still running and Parallel is supposed to continue only once all tasks have finished)